src/ccs/ir/circuit.rs

```diff
@@ -0,0 +1,579 @@
+                public_values[*idx - witness_len - 1] =
+                    *inputs.get(&(*step_idx, *signal_id)).unwrap();
+            }
+        }
+        self.assignments = witness_values;
+        self.public_inputs = public_values;
+    }
+
+    pub fn write_with_values(&mut self, inputs: &[F], witnesses: &[F]) {
+        assert_eq!(inputs.len(), self.l);
+        assert_eq!(witnesses.len(), self.n - self.l - 1);
+
+        self.public_inputs.clone_from(&inputs.to_owned());
+        self.assignments.clone_from(&witnesses.to_owned());
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Matrix<F> {
+    n: usize,
+    m: usize,
+    values: Vec<Vec<F>>,
+}
+
+impl<F: Field> Matrix<F> {
+    pub fn new(n: usize, m: usize) -> Self {
+        Self {
+            n,
+            m,
+            values: vec![vec![F::ZERO; n]; m],
+        }
+    }
+
+    pub fn write(&mut self, values: &[(usize, usize, F)]) {
+        for &(row, col, value) in values.iter() {
+            assert!(row < self.m);
+            assert!(col < self.n);
+            self.values[row][col] = value;
+        }
+    }
+
+    pub fn get(&self, row: usize, col: usize) -> F {
+        assert!(row < self.m);
+        assert!(col < self.n);
+        self.values[row][col]
+    }
+
+    pub fn reduce_rows(&mut self, m: usize) {
+        if m < self.m {
+            // self.values = self.values[0..m].to_owned();
+            self.values.truncate(m);
+            self.m = m;
+        }
+    }
+
+    pub fn size(&self) -> (usize, usize) {
+        (self.m, self.n)
+    }
+
+    pub fn values(&self) -> Vec<Vec<F>> {
+        self.values.clone()
+    }
+}
+
+impl<F: Field + From<u64> + Hash> Matrix<F> {
+    pub fn matrix_vector_product(&self, z: &[F]) -> Vec<F> {
+        (0..self.values.len())
+            .map(|i| self.hadamard_product_and_sum(i, z))
+            .collect()
+    }
+
+    fn hadamard_product_and_sum(&self, index: usize, z: &[F]) -> F {
+        assert!(index < self.values.len());
+        assert_eq!(self.values[index].len(), z.len());
+        hadamard_product(&self.values[index], z).iter().sum()
+    }
+}
+
+#[derive(Debug, Default)]
+pub struct CCSCircuit<F> {
+    pub n: usize,
+    pub m: usize,
+    pub l: usize,
+    pub t: usize,
+    pub q: usize,
+    pub d: usize,
+    pub matrices: Vec<Matrix<F>>,
+    pub selectors: SelectorsOffsetAndCoeff<F>,
+    pub constants: Vec<F>,
+}
+
+impl<F: Field + From<u64> + Hash> CCSCircuit<F> {
+    pub fn new() -> Self {
+        let matrices = Vec::new();
+        let selectors = Vec::new();
+        let constants = Vec::new();
+
+        Self {
+            n: 0,
+            m: 0,
+            l: 0,
+            t: 0,
+            q: 0,
+            d: 0,
+            matrices,
+            selectors,
+            constants,
+        }
+    }
+
+    pub fn public_num(&self) -> usize {
+        self.l
+    }
+
+    pub fn witness_num(&self) -> usize {
+        self.n - self.l - 1
+    }
+
+    pub fn write(
```

Comment:
@10to4 `write` is only necessary if this is modified, but it should not be modified as it is an IR. It is created by the compiler and then read by the backend